common/save-game.cpp

```diff
@@ -0,0 +1,147 @@
+	return std::find_if(this->users.begin(), this->users.end(), [id](proto::User user) { return user.id() == id; });
+}
+
+extern std::string LoadJSON(const FilePath& fname);
+
+bool SaveJSON(const FilePath& fname, std::string contents) {
+	std::fstream output(fname.GetNativePath(), std::ios::out);
+	if (!output.good())
+		throw std::runtime_error("can't open ." + fname.toString());
+
+	output.write(contents.c_str(), contents.length());
+	output.close();
+	return true;
+}
+
+bool SaveGame::Load(const FilePath _filepath) {
+	auto _log = spdlog::get("console_log");
+	this->filepath = _filepath;
+	if (_filepath.FileExists()) {
+		auto json_string = LoadJSON(this->filepath);
+		auto status = google::protobuf::util::JsonStringToMessage(json_string, &this->save);
+		if (status.ok()) {
+			this->LoadUsers();
+			this->LoadWorld();
+			return true;
+		}
+		_log->error("Failed to parse save data");
+		return false;
+	}
+
+	_log->error("File does not exist: {}", _filepath.FileName());
+	return false;
+}
+
+bool SaveGame::Save() { return this->Save(this->filepath); }
+
+bool SaveGame::Save(const FilePath _filepath) {
+	auto _log = spdlog::get("console_log");
+	if (_filepath.FileExists() || _filepath.isValidPath()) {
+		this->filepath = _filepath;
+		this->SaveUsers();
+		this->SaveWorld();
+
+		std::string json_string;
+
+		auto status = google::protobuf::util::MessageToJsonString(this->save, &json_string);
+
+		if (status.ok()) {
+			try {
+				SaveJSON(_filepath, json_string);
+			}
+			catch (std::runtime_error err) {
+				_log->error("Failed to save to file: {}", _filepath.FileName());
+			}
+			return true;
+		}
+		_log->error("Failed to serialize save data");
+		return false;
+	}
+	_log->error("File does not exist or path is invalid: {}", _filepath.toString());
+	return false;
+}
+
+UserList* SaveGame::GetUserList() { return &this->user_list; }
+
+void SaveGame::LoadUsers() {
+	auto users = this->save.mutable_users();
+	user_list.SetUsers(users->begin(), users->end());
+}
+
+void SaveGame::SaveUsers() {
+	auto users = this->save.mutable_users();
+	users->Clear();
+	for (const auto& user : *user_list.GetUsers()) {
+		auto save_user = users->Add();
+		save_user->CopyFrom(user);
+	}
+}
+
+void SaveGame::LoadWorld() {
+	auto _log = spdlog::get("console_log");
+	auto world = this->save.world();
+	for (int i = 0; i < world.entity_file_list_size(); i++) {
+		FilePath entity_filename = FilePath::GetAssetPath(world.entity_file_list(i));
+		if (entity_filename.isValidPath() && entity_filename.FileExists()) {
+			std::string json_string = LoadJSON(entity_filename);
+			proto::Entity entity;
+			auto status = google::protobuf::util::JsonStringToMessage(json_string, &entity);
+			if (status.ok()) {
+				EventSystem<EntityCreated>::Get()->Emit(
+						std::make_shared<EntityCreated>(EntityCreated{static_cast<eid>(entity.id()), entity}));
+			}
+			else {
+				_log->error("Failed to parse entity data");
```

Comment:
considering this is in the middle of a loop, it might be helpful to know *which* file it failed to parse.

Reply:
Yes, this case we do know which failed. I will update the error to include the filename.